adapter/avro/src/main/java/org/apache/arrow/adapter/avro/AvroFileReader.java

```diff
@@ -0,0 +1,374 @@
+    int validateMagic =
+        BinaryData.compareBytes(AVRO_MAGIC, 0, AVRO_MAGIC.length, magic, 0, AVRO_MAGIC.length);
+
+    if (validateMagic != 0) {
+      throw new RuntimeException("Invalid AVRO data file: The file is not an Avro file");
+    }
+
+    // Read the metadata map
+    for (long count = decoder.readMapStart(); count != 0; count = decoder.mapNext()) {
+
+      headerSize += zigzagSize(count);
+
+      for (long i = 0; i < count; i++) {
+
+        ByteBuffer keyBuffer = decoder.readBytes(null);
+        ByteBuffer valueBuffer = decoder.readBytes(null);
+
+        headerSize += zigzagSize(keyBuffer.remaining()) + keyBuffer.remaining();
+        headerSize += zigzagSize(valueBuffer.remaining()) + valueBuffer.remaining();
+
+        String key = new String(keyBuffer.array(), StandardCharsets.UTF_8);
+
+        // Handle header entries for schema and codec
+        if ("avro.schema".equals(key)) {
+          avroSchema = processSchema(valueBuffer);
+        } else if ("avro.codec".equals(key)) {
+          avroCodec = processCodec(valueBuffer);
+        }
+      }
+    }
+
+    // End of map marker
+    headerSize += 1;
+
+    // Sync marker denotes end of the header
+    decoder.readFixed(syncMarker);
+    headerSize += syncMarker.length;
+
+    // Schema must always be present
+    if (avroSchema == null) {
+      throw new RuntimeException("Invalid AVRO data file: Schema missing in file header");
+    }
+
+    // Prepare read config
+    this.dictionaries = new DictionaryProvider.MapDictionaryProvider();
+    AvroToArrowConfig config = new AvroToArrowConfig(allocator, 0, dictionaries, Set.of(), false);
+
+    // Calling this method will also populate the dictionary map
+    this.recordConsumer = AvroToArrowUtils.createCompositeConsumer(avroSchema, config);
+
+    // Initialize data vectors
+    List<FieldVector> vectors = new ArrayList<>(arrowSchema.getFields().size());
+    for (int i = 0; i < arrowSchema.getFields().size(); i++) {
+      FieldVector vector = recordConsumer.getConsumers().get(i).getVector();
+      vectors.add(vector);
+    }
+
+    // Initialize batch and schema
+    this.arrowBatch = new VectorSchemaRoot(vectors);
+    this.arrowSchema = arrowBatch.getSchema();
+
+    // First batch starts after the header
+    this.nextBatchPosition = headerSize;
+  }
+
+  private org.apache.avro.Schema processSchema(ByteBuffer buffer) throws IOException {
+
+    org.apache.avro.Schema.Parser parser = new org.apache.avro.Schema.Parser();
+
+    try (InputStream schemaStream = new ByteArrayInputStream(buffer.array())) {
+      return parser.parse(schemaStream);
+    }
+  }
+
+  private String processCodec(ByteBuffer buffer) {
+
+    if (buffer != null && buffer.remaining() > 0) {
+      return new String(buffer.array(), StandardCharsets.UTF_8);
+    } else {
+      return DataFileConstants.NULL_CODEC;
+    }
+  }
+
+  // Schema and VSR available after readHeader()
+  Schema getSchema() {
+    if (avroSchema == null) {
```

Comment:
Does it make sense to null-check the field itself (arrowSchema, not avroSchema)?

Should we use a Preconditions or Objects method?

Reply:
Or, given that this check repeats a lot - have a `checkStarted` method?